### PR TITLE
bug#10781. Remove Float range and Double range [ci: last-only]

### LINKED ISSUE #10781: Remove Float range and Double range
Ref https://github.com/scala/collection-strawman/pull/489
Ref scala/bug#8518
Ref scala/bug#9875
Ref scala/bug#9874
Ref scala/bug#8620
Ref scala/bug#8670

I tried to fix the Double range issue in https://github.com/scala/collection-strawman/pull/489, but ultimately I was not able to overcome the fact that addition in IEEE float doesn't behave in the way we need ranges to behave.

During the review @Ichoran suggested that we remove floating point ranges:

> Normally I don't like breaking changes, but in the case where there is no way to do the right thing except by not actually using the type that is stated, I think the better thing to do is remove the feature.

After thinking about it, I agree that it's the right thing to do since it doesn't work. If people want decimal ranges, BigDecimal range would behave in predictable way.

### PR DESCRIPTION
**Pull request by eed3si9n**
Fixes scala/bug#10781

At the basis of range is an assumption that summation works reliably. Unfortunately it is not true for IEEE floating point number where numbers are stored as approximate fraction of binary numbers. Natually, Double and Float ranges are completely broken. See numbers of issues such as scala/bug#8518, scala/bug#9875, scala/bug#9874, scala/bug#8620, scala/bug#8670, and scala/bug#10759.

I've attempted to fix the Double range in scala/collection-strawman#489 by faking it further using BigDecimal, but ultimately I was not able to overcome the fact that IEEE floats are not suited for ranges.

This removes both Float and Double ranges, as well as the fake `Integral[T]` instances that underpin their existence.

### CODE REVIEW

File: src/library/scala/math/Numeric.scala [hunk 1]
```diff
@@ -194,13 +184,12 @@ object Numeric {
     def rem(x: BigDecimal, y: BigDecimal): BigDecimal = x remainder y
   }
 
-  // For Double and BigDecimal we offer implicit Fractional objects, but also one
+  // For BigDecimal we offer implicit Fractional objects, but also one
```
Comment (NthPortal) [inline src/library/scala/math/Numeric.scala]: Should probably be singular "an implicit Fractional object" (admittedly, it wasn't grammatically correct in the first place)

File: src/library/scala/math/Numeric.scala [hunk 2]
```diff
@@ -168,10 +162,6 @@ object Numeric {
   trait DoubleIsFractional extends DoubleIsConflicted with Fractional[Double] {
```
Comment (NthPortal) [inline src/library/scala/math/Numeric.scala]: If `DoubleAsIfIntegral` is being removed, `DoubleIsConflicted` can probably be merged into `DoubleIsFractional`, as it's no longer conflicted (same for `FloatIsFractional`/`FloatIsConflicted`)

**Review by NthPortal — CHANGES_REQUESTED** [review]

File: src/library/scala/math/Numeric.scala (hunk 2 above)
Comment (eed3si9n) [reply src/library/scala/math/Numeric.scala]: Sounds good 👍

**Comment by szeiger** [comment]
This should target the 2.13.x-new-collections branch

**Comment by lrytz** [comment]
Yes, but let's just wait until the new collections are in 2.13.x - we only merge to 2.13.x-new-collections what's required to get the collections in 2.13.x.

**Comment by eed3si9n** [comment]
Rebased on top of 2.13.x-new-collections branch for testing, and addressed review comments.

**Comment by NthPortal** [comment]
I'm not disagreeing with @lrytz, but I think the target branch should be `2.13.x-new-collections` until that branch is merged, so that we can reason about what this PR actually changes (it doesn't change 30k lines)

**Comment by dwijnand** [comment]
here's the subset diff: https://github.com/scala/scala/compare/2.13.x-new-collections...eed3si9n:wip/float-range

**Comment by eed3si9n** [comment]
Rebased and squashed.

**Comment by SethTisue** [comment]
why remove immediately, instead of deprecate-in-2.13, remove-in-2.14?

**Review by NthPortal — APPROVED** [review]

**Comment by eed3si9n** [comment]
> why remove immediately, instead of deprecate-in-2.13, remove-in-2.14?

Deprecation is a useful tool to spread the cost of code migration targeting code base that was once in vogue, but no longer considered good, like cars without seatbelts and smoking cigarette in bars. It takes a while for all cars to catch up to the safety standards. That would be our procedure syntax and view bounds.

Floating point ranges, on the other hand, are as broken as cell phones that can catch fire on their own, or car wheels coming off on highways. If someone is (somehow) using it it in 2.12, we should advise them to stop too.

To quote @Ichoran in https://github.com/scala/collection-strawman/pull/489#issuecomment-370528213:

> Is it possible to drop numeric ranges that pretend to be for `Float` and `Double`?  These are just bad news due to numeric imprecision.  If you say something like `0.1 to 0.7 by (0.1 + 0.2)` you would _think_ you would have three steps, but you either won't, or you will but other code will be randomly inaccurate (because you rounded things enough to make the above work). 
> (snip)
>
> Normally I don't like breaking changes, but in the case where there is no way to do the right thing except by not actually using the type that is stated, I think the better thing to do is remove the feature.

**Comment by lrytz** [comment]
If we remove this in 2.13.x (which is fine with me), let's deprecate it ASAP in 2.12 (we can still make it for 2.12.6).

**Comment by eed3si9n** [comment]
@lrytz Here's a PR for deprecation in 2.12.x - https://github.com/scala/scala/pull/6550

**Comment by SethTisue** [comment]
I had previously expressed misgivings about removal at https://github.com/scala/collection-strawman/pull/489#issuecomment-370627602, but I think I'm happy to go along with the near-consensus that has developed on it.

**Comment by SethTisue** [comment]
the deprecation PR is merged for 2.12.6.

this PR needs a rebase.

**Comment by eed3si9n** [comment]
@SethTisue Done.

**Comment by dwijnand** [comment]
The deprecation PR merged in 2.12.x was merged into 2.13.x (#6575), meaning this PR needs another rebase. Thank you for playing Code Branches (tm) with scala/scala.

**Comment by eed3si9n** [comment]
Rebased on top of 2.13.x.

**Comment by NthPortal** [comment]
should we merge this now?

**Review by lrytz — APPROVED** [review]

**Comment by lrytz** [comment]
Thanks @eed3si9n, @NthPortal!

**Comment by Ichoran** [comment]
@eed3si9n @NthPortal @SethTisue @som-snytt - What if we change the API?

The whole problem is that we can't hit the endpoints accurately by specifying a step.  So what if we abandon the idea of specifying a precise step?

Suppose that `0.1 to 0.7 into 7` will give seven elements that are roughly 0.1, 0.2, ..., 0.7, which is about as much as you can expect from `Double` anyway.  Now there isn't any problem.  We don't even have to let `until` work any more.  We could also introduce `0.1 to 0.7 every 0.1` with the meaning that `0.1` and `0.7` are _guaranteed_ to be the endpoints, and it will equally space the points in between as close to `0.1` as it can.  Again, no problems with randomly wrong endpoints, because you're required to specify the endpoints.  And the intermediates will be as close as anything ever is with `Double` arithmetic.

WDYT?

**Comment by Ichoran** [comment]
Bikeshed: maybe `spaced` is better than `every` for indicating that the spacing is a rough guideline.

**Comment by som-snytt** [comment]
On https://github.com/scala/scala/pull/6803 I was going to give a shot at something like just ensuring that the `mapRange` is convenient to use. That is, I don't see the use case as specifying a double step, but just to make it hard to do the wrong thing with range.double with bigdecimal step. I noticed that `BigDecimal(.1 * 3)` is still an open door. I was going to ask on the forum whether it's been considered to let `.1 * 3 : BigDecimal` do the obvious. Someone asked about it on the other thread. Also, the `to` and `by` could require `BigDecimal`.

Your idea of doing something mathy with a divided interval is also interesting as an extra credit question.

The `dice` operator, `|/|`. As in slice and dicer from Ronco.

**Comment by eed3si9n** [comment]
If we _really_ want some `Double` operations, I'd say we look into @non's Spire. I remember Intervals talk from nescala a while back - http://plastic-idolatry.com/erik/nescala2015.pdf

**Comment by NthPortal** [comment]
Honestly, I'm not sure that such an API would be *that* useful, and would still expect confusion

**Pull request merged**